### PR TITLE
Improve IResourceActivity resource abstraction

### PR DESCRIPTION
**Pull request by tobous**
This PR is a big refactoring starting to phase out the usage of `SPath`. `SPath` was meant as a transport wrapper for resource objects. But, in its current usage, it often also functions as a resource identifier outside of the transport logic. Such usages only add unnecessary bulk as the described resource should be used directly instead if the marshaling/unmarshaling of the resource is not of interest.

This issue seems to stem from the activities allowing this wrapper object to be accessed from the outside. To avoid this issue in the future, the adjusted logic no longer allows callers to access the resource wrapper object but instead just offers access to the resource directly.

In the process, the `IResourceActivity` interface is also made generic to allow easier access to the right view on the resource.

Introduces `ResourceTransportWrapper` as a generic transport wrapper to replace `SPath`.

The remaining usage of `SPath` will be removed in follow-up PR(s).

## Reviewing This PR

This PR contains 6 major refactoring steps:
1. The introduction of the `IResourceActivity.getResource` method and adjusting the resource activity CTORs to use resource objects instead of `SPath` objects.
2. Removing the usage of `IResourceActivity.getPath()` from the core.
3. Removing the usage of `IResourceActivity.getPath()` from the Eclipse implementation.
4. Removing the usage of `IResourceActivity.getPath()` from the IntelliJ implementation.
5. Removing the usage of `IResourceActivity.getPath()` from the Server implementation.
6. Making `IResourceActivity` generic and introducing `ResourceTransportWrapper` as a generic replacement of `SPath` for the resource transport in resource activities.

The commits for each phase are listed below. As this is a big refactoring that required some ugly temporary solutions, I would suggest reviewing this PR by only looking at the final state (i.e. all commits combined).

Most of the changes are simple replacements of the usage of `SPath` with `IResource`, `IFile`, or `IFolder`. Major new logic is only introduced with the commit `[INTERNAL][CORE] Introduce ResourceTransportWrapper`, adding `ResourceTransportWrapper`, `ResourceTransportWrapperConverter`, and `ResourceTransportWrapperConverterTest`.

This PR is somewhat inconsistent on how much of the surrounding logic is refactored to use resource instead of `SPath`. As I wanted to keep this first halve of the refactoring as small as possible, I tried to do as few adjustments as possible unless the usage of `SPath` could completely be removed from a class/method. As a result, this PR adds a lot of new instantiations of `new SPath(IResource)` in places where other logic still requires `SPath` objects. These calls will be cleaned up once the surrounding logic has been cleaned up.

### Step 1 - Use resource in `IResourceActivity` CTORs
<br>

<details><summary><b>[API] Add IResourceActivity.getResource()</b></summary>
<br>

Adds the method IResourceActivity.getResource(). The method will be
used replace the usage of getPath() in upcoming changes. This will be
done to avoid the unnecessary usage of SPath outside of the core as it
is only meant as a wrapper class to marshal and unmarshal resources.

</details>

<details><summary><b>[INTERNAL] Require IFile as CTOR parameter for file activities</b></summary>
<br>

Adjusts the constructor for FileActivity and TargetedFileActivity to
accept IFile objects instead of SPath objects. Adjusts all logic calling
the constructors accordingly.

</details>

<b>[INTERNAL] Require IFile as CTOR parameter for checksum activities</b>

<b>[INTERNAL] Require IFile as CTOR parameter for file deletion ACK</b>

<b>[INTERNAL] Require IFile as CTOR parameter for editor activities</b>

<b>[INTERNAL] Require IFolder as CTOR parameter for folder activities</b>

<details><summary><b>[JUNIT] Introduce helper method to create resource backed SPath mocks</b></summary>
<br>

Introduces SarosMocks.mockResourceBackedSPath() which can be used to
obtain an SPath object backed by mocked resource objects. As a result,
the mock allows for cyclic conversion between SPath and IResource
objects.

</details>

<b>[INTERNAL] Require IFile as CTOR parameter for jupiter activities</b>

<b>[INTERNAL] Require IFile as CTOR parameter for selection activities</b>

<b>[INTERNAL] Require IFile as CTOR parameter for viewport activities</b>

<b>[INTERNAL] Require IFile as CTOR parameter for text edit activities</b>

<b>[INTERNAL] Require IFile to transform operation to text edit activity</b>

<b>[INTERNAL] Require IFile as CTOR parameter for checksum error activities</b>

### Step 2 - Remove usage of `IResourceActivity.getPath()` from core
<br>

<b>[INTERNAL][CORE] Remove usage of SPath from ActivityOptimizer</b>

<b>[INTERNAL][CORE] Remove usage of SPath from ActivityQueuer</b>

<b>[INTERNAL] Remove usage of SPath from ConsistentcyWatchdogClient</b>

<b>[INTERNAL][CORE] Remove usage of SPath from DocumentChecksum</b>

<b>[INTERNAL][CORE] Remove usage of SPath from ConsistencyWatchdogServer</b>

<details><summary><b>[INTERNAL][CORE] Remove direct usage of SPath from FollowModeManager</b></summary>
<br>

The class is still used, but only to use method that were not yet
refactored.

</details>

<b>[INTERNAL][CORE] Remove usage of SPath from JupiterClient</b>

<details><summary><b>[INTERNAL][CORE] Remove usage of SPath from JupiterServer</b></summary>
<br>

Subsequently removes usage of SPath from JupiterDocumentServer.

</details>

<b>[INTERNAL][CORE] Remove usage of SPath from ResourceActivityFilter</b>

<b>[INTERNAL][CORE] Remove usage of SPath from SarosSession</b>

<b>[INTERNAL][CORE] Remove usage of SPath from UserEditorState</b>

<b>[INTERNAL][CORE] Remove usage of SPath from ActivityHandler</b>

### Step 3 - Remove usage of `IResourceActivity.getPath()` from Eclipse implementation
<br>

<b>[INTERNAL][E] Remove usage of SPath from EditorActivityDelayer</b>

<b>[INTERNAL][E] Remove usage of SPath from FileActivityConsumer</b>

<b>[INTERNAL][E] Remove usage of SPath from FolderActivityConsumer</b>

<details><summary><b>[INTERNAL][E] Remove usage of SPath from RemoteWriteAccessManager</b></summary>
<br>

Also reduces the visibility the contained variables and methods to the
minimal required visibility.

</details>

<b>[INTERNAL][E] Remove usage of SPath from SelectionCollector</b>

<details><summary><b>[INTERNAL][E] Remove usage of SPath from UndoManager</b></summary>
<br>

Subsequently removes the usage of SPath from OperationHistory.

</details>

<details><summary><b>[INTERNAL][E] Remove usage of getPath() from EditorManager</b></summary>
<br>

Removes the usage of IResourceActivity.getPath() by replacing it with
getResource().

The rest of the usages of SPath have to be removed in a separate
refactoring.

</details>

### Step 4 - Remove usage of `IResourceActivity.getPath()` from IntelliJ implementation
<br>

<details><summary><b>[INTERNAL][I] Remove usage of getPath() from SharedResourcesManager</b></summary>
<br>

Removes the usage of IResourceActivity.getPath() by replacing it with
getResource().

The rest of the usages of SPath have to be removed in a separate
refactoring.

</details>

<details><summary><b>[INTERNAL][I] Remove usage of getPath() from EditorManager</b></summary>
<br>

Removes the usage of IResourceActivity.getPath() by replacing it with
getResource().

The rest of the usages of SPath have to be removed in a separate
refactoring.

</details>

### Step 5 - Remove usage of `IResourceActivity.getPath()` from Server implementation
<br>

<b>[INTERNAL][S] Remove usage of SPath from FileActivityExecutor</b>

<b>[INTERNAL][S] Remove usage of SPath from FolderActivityExecutor</b>

<details><summary><b>[INTERNAL][S] Remove usage of getPath() from ServerEditorManager</b></summary>
<br>

Removes the usage of IResourceActivity.getPath() by replacing it with
getResource().

The rest of the usages of SPath have to be removed in a separate
refactoring.

</details>

### Step 6 - Make `IResourceActivity` generic and replace `SPath` with `ResourceTransportWrapper` as resource transport wrapper
<br>

<b>[API][CORE] Remove IResourceActivity.getPath()</b>

<details><summary><b>[INTERNAL][CORE] Introduce ResourceTransportWrapper</b></summary>
<br>

Introduces ResourceTransportWrapper, a simpliefied, generic version of
SPath. The class will be used to replace the usage of SPath for resource
activities. It is meant only as an internal class and should not be
passed out of the activity object.

Introduces ResourceTransportWrapperConverter to marshal and unmarshar
ResourceTransportWrapper objects.

Both ResourceTransportWrapperConverter and
ResourceTransportWrapperConverterTest were closely modeleted after
SPathConverter and SPathConverter respectively.

</details>

<details><summary><b>[INTERNAL][CORE] Use ResourceTransportWrapper for resource activities</b></summary>
<br>

Replaces the usage of SPath in IResourceActivity implementations with
using ResourceTransportWrapper.

Makes the interface IResourceActivity generic to allow for a better
usage. The implementing classes now declare which resource type they are
interested in and will be able to directly access the contained resource
through the correct resource type interface.

Moves the implementation of IResourceActivity.getResource() to
AbstractResourceActivity. This allows the new implementation to never
leak the resource wrapper.

Updates ChecksumErrorActivity to also use ResourceTransportWrapper
instead of SPath.

Marks SPath as deprecated.

Removes SPathConverter as is now no longer needed.

</details>

<details><summary><b>[JUNIT] Clean up test cases adjusted during SPath removal</b></summary>
<br>

Cleans up some superfluous mocking logic that is no longer needed with
the introduction of ResourceTransportWrapper.

Removes the method SarosMocks.mockResourceBackedSPath() and replaces it
with mockFile().

</details>

<b>[FIX] Avoid raw use of parameterized interface IResourceActivity</b>

### CODE REVIEW

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 14
           

Complexity increasing per file
==============================
- core/src/saros/misc/xstream/ResourceTransportWrapperConverter.java  5
- core/src/saros/activities/FileActivity.java  2
- core/test/junit/saros/misc/xstream/ResourceTransportWrapperConverterTest.java  2
- core/src/saros/activities/ResourceTransportWrapper.java  4
- core/src/saros/concurrent/management/ResourceActivityFilter.java  1
- core/test/junit/saros/session/internal/ActivityHandlerTest.java  2
         

Complexity decreasing per file
==============================
+ core/src/saros/activities/AbstractResourceActivity.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5401736&bid=17708116)

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 8
           

Complexity increasing per file
==============================
- core/src/saros/activities/ResourceTransportWrapper.java  4
- core/src/saros/misc/xstream/ResourceTransportWrapperConverter.java  5
- core/test/junit/saros/session/internal/ActivityHandlerTest.java  2
- core/src/saros/concurrent/management/ResourceActivityFilter.java  1
- core/test/junit/saros/misc/xstream/ResourceTransportWrapperConverterTest.java  2
- core/src/saros/activities/FileActivity.java  2
         

Complexity decreasing per file
==============================
+ core/src/saros/activities/AbstractResourceActivity.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5401736&bid=17708116)

**Comment by tobous** [comment]
I initially planned on splitting this PR into multiple smaller PRs along what are now labeled as the different steps. But, as mentioned in the PR message, the refactoring contains many ugly in-between states that are later cleaned up as part of the refactoring, so I decided to keep it as one PR and would suggest to review all changes at once. This will lead to a *very* long list of changes, but most of them are simple replacements of the usage of `SPath` with `IResource`, `IFile`, or `IFolder`.

If you see this differently, I can still split this into different PRs.

File: core/test/junit/saros/editor/remote/UserEditorStateTest.java [hunk 1]
```diff
@@ -19,34 +19,28 @@
...
+  private static IFile fileA;
+  private static IFile fileB;
 
   private static User source;
 
   @BeforeClass
   public static void prepare() {
     /* Mocks */
-    IPathFactory pathFactory = EasyMock.createMock(IPathFactory.class);
-    IPath pathMockA = mockPath("/foo/src/Main.java", pathFactory);
```
Comment (tobous) [inline core/test/junit/saros/editor/remote/UserEditorStateTest.java:41]: Remove now unused method `mockPath(...)`

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 8
           

Complexity increasing per file
==============================
- core/src/saros/activities/FileActivity.java  2
- core/src/saros/misc/xstream/ResourceTransportWrapperConverter.java  5
- core/src/saros/activities/ResourceTransportWrapper.java  4
- core/src/saros/concurrent/management/ResourceActivityFilter.java  1
- core/test/junit/saros/session/internal/ActivityHandlerTest.java  2
- core/test/junit/saros/misc/xstream/ResourceTransportWrapperConverterTest.java  2
         

Complexity decreasing per file
==============================
+ core/src/saros/activities/AbstractResourceActivity.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5401736&bid=17708116)

**Comment by tobous** [comment]
Rebased onto current base branch resolving some minor merge issues.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 8
           

Complexity increasing per file
==============================
- core/test/junit/saros/misc/xstream/ResourceTransportWrapperConverterTest.java  2
- core/test/junit/saros/session/internal/ActivityHandlerTest.java  2
- core/src/saros/misc/xstream/ResourceTransportWrapperConverter.java  5
- core/src/saros/activities/ResourceTransportWrapper.java  4
- core/src/saros/activities/FileActivity.java  2
- core/src/saros/concurrent/management/ResourceActivityFilter.java  1
         

Complexity decreasing per file
==============================
+ core/src/saros/activities/AbstractResourceActivity.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5401736&bid=17708116)

File: core/test/junit/saros/concurrent/jupiter/test/puzzles/SimpleJupiterDocumentTest.java [hunk 2]
```diff
@@ -1,38 +1,38 @@
...
 
 /**
  * This test class represents the local execution of document changes and appropriate jupiter
  * operations.
  */
 public class SimpleJupiterDocumentTest extends JupiterTestCase {
 
+  private final IFile fileMock;
+
+  public SimpleJupiterDocumentTest() {
```
Comment (m273d15) [inline core/test/junit/saros/concurrent/jupiter/test/puzzles/SimpleJupiterDocumentTest.java:26]: Could be in a `@Before` method

File: core/src/saros/communication/extensions/ActivitiesExtension.java [hunk 3]
```diff
@@ -136,8 +136,8 @@ private Provider() {
           SplitOperation.class,
           TimestampOperation.class,
 
-          // SPATH
-          SPath.class,
+          // Resource transport wrapper
```
[the remark this answers is not in the document]
Comment (tobous) [reply core/src/saros/communication/extensions/ActivitiesExtension.java:139]: It was there beforehand, so I kept it. It is useful to separate the activities from other classes.

File: core/test/junit/saros/concurrent/jupiter/test/puzzles/SimpleJupiterDocumentTest.java (hunk 2 above)
Comment (tobous) [reply core/test/junit/saros/concurrent/jupiter/test/puzzles/SimpleJupiterDocumentTest.java:26]: Done.

File: core/src/saros/activities/FileActivity.java [hunk 4]
```diff
@@ -85,21 +86,26 @@ public FileActivity(
     }
 
     this.type = type;
-    this.oldPath = oldPath;
+    this.oldFile = oldFile != null ? new ResourceTransportWrapper<>(oldFile) : null;
```
[the remark this answers is not in the document]
Comment (tobous) [reply core/src/saros/activities/FileActivity.java:89]: Renamed to `oldFileWrapper`.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 8
           

Complexity increasing per file
==============================
- core/test/junit/saros/misc/xstream/ResourceTransportWrapperConverterTest.java  2
- core/test/junit/saros/session/internal/ActivityHandlerTest.java  2
- core/src/saros/concurrent/management/ResourceActivityFilter.java  1
- core/src/saros/activities/FileActivity.java  2
- core/src/saros/activities/ResourceTransportWrapper.java  4
- core/src/saros/misc/xstream/ResourceTransportWrapperConverter.java  5
         

Complexity decreasing per file
==============================
+ core/src/saros/activities/AbstractResourceActivity.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5401736&bid=17708116)

**Comment by tobous** [comment]
Rebased onto current master without any interaction.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 6
           

Complexity increasing per file
==============================
- core/test/junit/saros/misc/xstream/ResourceTransportWrapperConverterTest.java  2
- core/src/saros/concurrent/management/ResourceActivityFilter.java  1
- core/src/saros/misc/xstream/ResourceTransportWrapperConverter.java  5
- core/test/junit/saros/session/internal/ActivityHandlerTest.java  2
- core/src/saros/activities/ResourceTransportWrapper.java  4
- core/src/saros/activities/FileActivity.java  2
         

Complexity decreasing per file
==============================
+ core/src/saros/activities/AbstractResourceActivity.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5401736&bid=17708116)

**Pull request merged**